### PR TITLE
Chunked encoding

### PR DESCRIPTION
**Pull request by graebm**
Originally reviewed here:
https://github.com/awslabs/aws-c-http/pull/237

Authored by @czakian, but I'm merging to master so we can get our CI to run on it. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/http/private/h1_encoder.h [hunk 1]
```diff
@@ -15,9 +15,38 @@
...
+};
+
+enum aws_h1_encoder_body_stream_state {
+    AWS_H1_ENCODER_STATE_CHUNK_INIT,
+    AWS_H1_ENCODER_STATE_CHUNK_LINE,
+    AWS_H1_ENCODER_STATE_CHUNK_PAYLOAD,
+    AWS_H1_ENCODER_STATE_CHUNK_END,
+    AWS_H1_ENCODER_STATE_CHUNK_TERMINATED,
+};
+#define MAX_ASCII_HEX_CHUNK_STR_SIZE sizeof(size_t) * 2 + 1
```
Comment (JonathanHenson) [inline include/aws/http/private/h1_encoder.h:47]: NIT: either put in parens or make it a constant

File: include/aws/http/private/h1_stream.h [hunk 2]
```diff
@@ -24,6 +24,9 @@ struct aws_h1_stream {
 
     struct aws_linked_list_node node;
 
+    /* Linked list of stream to be encoded on the same connection via transfer encoding chunked */
```
Comment (JonathanHenson) [inline include/aws/http/private/h1_stream.h:27]: THIS is the linked list?

File: source/h1_encoder.c [hunk 3]
```diff
@@ -52,18 +53,41 @@ static int s_scan_outgoing_headers(
...
-                if (encoder_message->content_length > 0) {
-                    has_body_headers = true;
+            } break;
+            case AWS_HTTP_HEADER_TRANSFER_ENCODING: {
+                has_transfer_encoding_header = true;
+                if (0 == header.value.len) {
+                    AWS_LOGF_ERROR(AWS_LS_HTTP_STREAM, "id=static: Transfer-Encoding must include a valid value");
+                    return aws_raise_error(AWS_ERROR_HTTP_INVALID_HEADER_VALUE);
+                }
+                struct aws_byte_cursor chunked = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("chunked");
```
Comment (JonathanHenson) [inline source/h1_encoder.c:69]: not sure why we don't hoist that up out of the function scope. Compiler prolly doesn't care but seems cleaner.

File: source/h1_encoder.c [hunk 4]
```diff
@@ -319,6 +371,409 @@ int aws_h1_encoder_start_message(
...
+    char crlf[CRLF_SIZE] = {'\r', '\n'};
+    struct aws_byte_cursor crlf_cursor = aws_byte_cursor_from_array(&crlf, AWS_ARRAY_SIZE(crlf));
+    return aws_byte_buf_write_from_whole_cursor(dst, crlf_cursor);
+}
+
+bool aws_write_chunk_size(struct aws_byte_buf *dst, size_t chunk_size) {
+    AWS_PRECONDITION(dst);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    AWS_PRECONDITION((dst->capacity - dst->len) >= MAX_ASCII_HEX_CHUNK_STR_SIZE);
+    char ascii_hex_chunk_size_str[MAX_ASCII_HEX_CHUNK_STR_SIZE];
```
Comment (JonathanHenson) [inline source/h1_encoder.c:463]: = {0} works just as well here.

File: source/h1_stream.c [hunk 5]
```diff
@@ -32,10 +33,93 @@ static void s_stream_update_window(struct aws_http_stream *stream, size_t increm
...
 
+static int s_body_chunks_init(struct aws_h1_stream *stream) {
+    AWS_PRECONDITION(stream);
+    aws_linked_list_init(&stream->body_chunks.chunk_list);
+    stream->body_chunks.current_chunk = NULL;
+    stream->body_chunks.paused = false;
+    return aws_mutex_init(&stream->body_chunks.lock);
+}
+
+void s_clean_up_body_chunk(struct aws_http1_stream_chunk **chunk) {
```
Comment (JonathanHenson) [inline source/h1_stream.c:44]: static

**Comment by graebm** [comment]
> LGTM, b.t.w what happens if gzip is used?

The user would need to gzip the data themselves. I'll add a comment.

File: include/aws/http/private/h1_stream.h (hunk 2 above)
Comment (graebm) [reply include/aws/http/private/h1_stream.h:27]: rephrased

File: source/h1_encoder.c (hunk 3 above)
Comment (graebm) [reply source/h1_encoder.c:69]: made this look like the other equality checks in the file

File: source/h1_stream.c (hunk 5 above)
Comment (graebm) [reply source/h1_stream.c:44]: I regex-found a bunch more non-static "s_" functions all over the library and fixed them

**Pull request merged**